File: app/src/main/java/me/zcraig/threaddownloader/MainActivity.java
package me.zcraig.threaddownloader;

import android.Manifest;
import android.content.pm.PackageManager;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import androidx.work.Data;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkManager;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, 1);
        }
        setContentView(R.layout.activity_main);
        Button download = findViewById(R.id.downloadButton);
        download.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Data data = new Data.Builder().putString("url", getLinkText()).putString("selector", getSelectorText()).build();
                OneTimeWorkRequest req = new OneTimeWorkRequest.Builder(DownloadTask.class).setInputData(data).build();
                WorkManager.getInstance().enqueue(req);
            }
        });
    }

    public String getLinkText(){
        return ((EditText) this.findViewById(R.id.urlText)).getText().toString();
    }

    public String getSelectorText(){
        return ((EditText) this.findViewById(R.id.selectorText)).getText().toString();
    }
}

File: app/src/main/java/me/zcraig/scraper/MainActivity.kt
package me.zcraig.scraper

import android.Manifest
import android.content.ClipboardManager
import android.content.pm.PackageManager
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.EditText
import androidx.work.Data
import androidx.work.OneTimeWorkRequest
import androidx.work.WorkManager

class MainActivity : AppCompatActivity() {

    private val linkTextField: EditText
        get() = (this.findViewById<View>(R.id.urlText) as EditText)

    private val selectorTextField: EditText
        get() = (this.findViewById<View>(R.id.selectorText) as EditText)

    private val linkText: String
        get() = linkTextField.text.toString()

    private val selectorText: String
        get() = selectorTextField.text.toString()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), 1)
        }
        setContentView(R.layout.activity_main)

        if (selectorText.isEmpty()) {
            selectorTextField.setText("a.fileThumb");
        }

        val download = findViewById<Button>(R.id.downloadButton)
        download.setOnClickListener {
            if (linkText.isEmpty()) {
                linkTextField.selectAll();
                val clippy = getSystemService(CLIPBOARD_SERVICE) as ClipboardManager;
                if (clippy.hasPrimaryClip()) {
                    val item = clippy.primaryClip!!.getItemAt(0);
                    val clipboardUrl = item!!.text;
                    Log.i(getString(R.string.app_name), "Clipboard URL: $clipboardUrl")
                    linkTextField.setText(clipboardUrl)
                } else {
                    Log.i(getString(R.string.app_name), "No clipboard contents :(")
                }
            }
            val data = Data.Builder().putString("url", linkText).putString("selector", selectorText).build()
            val req = OneTimeWorkRequest.Builder(DownloadTask::class.java).setInputData(data).build()
            WorkManager.getInstance().enqueue(req)
        }
    }
}
